Source/Core/Heap.js

```diff
@@ -0,0 +1,200 @@
+/*global define*/
+define([
+        './Check',
+        './defaultValue',
+        './defined',
+        './defineProperties'
+    ], function(
+        Check,
+        defaultValue,
+        defined,
+        defineProperties) {
+    'use strict';
+
+    /**
+     * Array implementation of a heap.
+     *
+     * @alias Heap
+     * @constructor
+     * @private
+     *
+     * @param {Function} comparator The comparator to use for the heap. If comparator(a, b) is less than 0, sort a to a lower index than b, otherwise sort to a higher index.
+     */
+    function Heap(comparator) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('comparator', comparator);
+        //>>includeEnd('debug');
+
+        this._comparator = comparator;
+        this._array = [];
+        this._length = 0;
+        this._maximumLength = undefined;
+    }
+
+    defineProperties(Heap.prototype, {
+        /**
+         * Gets the length of the heap.
+         *
+         * @memberof Heap.prototype
+         *
+         * @type {Number}
+         * @readonly
+         */
+        length : {
+            get : function() {
+                return this._length;
+            }
+        },
+
+        /**
+         * Gets the internal array.
+         *
+         * @memberof Heap.prototype
+         *
+         * @type {Array}
+         * @readonly
+         */
+        internalArray : {
```

Comment:
Kind of odd to expose this, but since it's a private class, I guess I'm okay with it.